source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -50,11 +54,139 @@ void Filter::onRead() {
+    if (!proxy_protocol_header_.value().remote_address_->ip()->isUnicastAddress() ||
+        !proxy_protocol_header_.value().local_address_->ip()->isUnicastAddress()) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+
+    // Only set the local address if it really changed, and mark it as address being restored.
+    if (*proxy_protocol_header_.value().local_address_ != *socket.localAddress()) {
+      socket.setLocalAddress(proxy_protocol_header_.value().local_address_, true);
+    }
+    socket.setRemoteAddress(proxy_protocol_header_.value().remote_address_);
+  }
+
+  // Release the file event so that we do not interfere with the connection read events.
+  file_event_.reset();
+  cb_->continueFilterChain(true);
+}
+
+size_t Filter::lenV2Address(char* buf) {
+  const uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+  const int ver_cmd = buf[PROXY_PROTO_V2_SIGNATURE_LEN];
+  size_t len;
+
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_LOCAL) {
+    // According to the spec there is no address encoded, len=0, and we must ignore
+    return 0;
+  }
+
+  switch ((proto_family & 0xf0) >> 4) {
+  case PROXY_PROTO_V2_AF_INET:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET;
+    break;
+  case PROXY_PROTO_V2_AF_INET6:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET6;
+    break;
+  default:
+    throw EnvoyException("Unsupported V2 proxy protocol address family");
+  }
+  return len;
+}
+
+void Filter::parseV2Header(char* buf) {
+  const int ver_cmd = buf[PROXY_PROTO_V2_SIGNATURE_LEN];
+  uint8_t upper_byte = buf[PROXY_PROTO_V2_HEADER_LEN - 2];
+  uint8_t lower_byte = buf[PROXY_PROTO_V2_HEADER_LEN - 1];
+  size_t hdr_addr_len = (upper_byte << 8) + lower_byte;
+
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_LOCAL) {
+    // This is locally-initiated, e.g. health-check, and should not override remote address
+    proxy_protocol_header_.emplace(WireHeader{hdr_addr_len});
+    return;
+  }
+
+  // Only do connections on behalf of another user, not internally-driven health-checks. If
+  // its not on behalf of someone, or its not AF_INET{6} / STREAM/DGRAM, ignore and
+  /// use the real-remote info
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_ONBEHALF_OF) {
+    uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+    if (((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_STREAM) ||
+        ((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_DGRAM)) {
+      if (((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET) {
+        typedef struct {
+          uint32_t src_addr;
+          uint32_t dst_addr;
+          uint16_t src_port;
+          uint16_t dst_port;
+        } __attribute__((packed)) pp_ipv4_addr;
+        pp_ipv4_addr* v4;
+        v4 = reinterpret_cast<pp_ipv4_addr*>(&buf[PROXY_PROTO_V2_HEADER_LEN]);
+        sockaddr_in ra4, la4;
+        ra4.sin_family = AF_INET;
+        ra4.sin_port = v4->src_port;
+        ra4.sin_addr.s_addr = (v4->src_addr);
```

Comment:
You can still remove parens here, and on other similar lines in this function (152, 153, 170).